server/src/main/java/org/jboss/as/server/BootstrapImpl.java

```diff
@@ -274,5 +282,82 @@ public void handleTermination(Info info) {
+                final SuspendController suspendController = getSuspendController(sc);
+                if (suspendController != null) {
+                    long timeout = getSuspendTimeout();
+                    final CountDownLatch suspendLatch = new CountDownLatch(1);
+                    OperationListener listener = new OperationListener() {
+                        @Override
+                        public void suspendStarted() {
+
+                        }
+
+                        @Override
+                        public void complete() {
+                            suspendController.removeListener(this);
+                            suspendLatch.countDown();
+                        }
+
+                        @Override
+                        public void cancelled() {
+                            suspendController.removeListener(this);
+                            suspendLatch.countDown();
+                        }
+
+                        @Override
+                        public void timeout() {
+                            suspendController.removeListener(this);
+                            suspendLatch.countDown();
+                        }
+                    };
+                    suspendController.addListener(listener);
+                    suspendController.suspend(timeout);
+                    if (timeout > 0) {
+                        // The latch should trip within 'timeout' but if necessary we'll wait
+                        // 500 ms longer for it in the off chance a gc or something delays things
+                        suspendLatch.await(timeout + 500, TimeUnit.MILLISECONDS);
+                    } else if (timeout < 0) {
+                        suspendLatch.await();
+                    } // else 0 means we don't wait for tasks to finish
+                }
+            } catch (InterruptedException ie) {
+                Thread.currentThread().interrupt();
+            }
+            catch (Throwable t) {
+                // ignored
+            }
+        }
+
+        private static SuspendController getSuspendController(ServiceContainer sc) {
+            SuspendController result = null;
+            if (sc != null && !sc.isShutdownComplete()) {
+                final ServiceController serviceController = sc.getService(SuspendController.SERVICE_NAME);
+                if (serviceController != null && serviceController.getState() == ServiceController.State.UP) {
+                    result = (SuspendController) serviceController.getValue();
+                }
+            }
+            return result;
+        }
+
+        private static long getSuspendTimeout() {
+            long result = 0;
+            String timeoutString = System.getProperty(SIGTERM_SUSPEND_TIMEOUT_PROP);
+            if (timeoutString != null && timeoutString.length() > 0) {
+                try {
+                    int max = Integer.decode(timeoutString);
+                    result = max > 0 ? max * 1000 : max;
```

Comment:
Should we care about overflow?  Cast either max or 1000 to type long as `result` is long